### DISCUSSION TITLE
How do I export from one machine and integrate on another?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#19410] [Q&A]
I have read about *export* (https://devblogs.microsoft.com/cppblog/vcpkg-introducing-export-command/) but I do not wish to have developers install a .nuget. I want them to use vcpkg mirroring our build-server, but not have to build everything themselves. This way their setup better mirrors the build machine.

Reading https://vcpkg.readthedocs.io/en/latest/users/integration/#export-command it mentions:

> Each of these have the same layout, which mimics the layout of a full vcpkg:
> - installed\ contains the installed package files

Does that mean I can use the following process:

1. Build-server: Run _vcpkg install_ , this will be used in our builds
2. Build-server: Run _vcpkg export_ to generate ZIP/raw folder <vcpkg-export-dir>
3. My PC: install clean vcpkg
4. My PC: copy <build-server>\<vcpkg-export-dir> => .\vcpkg\installed
5. My PC: _vcpkg integrate install_

I have run into issues before trying to hack file-structure so I'd like to be clear if this will work properly... if a developer then wants to install a new port on their PC, or rebuild one that was exported, will this work? Our full _vcpkg install_ now takes several hours so every time I miss something and redo it is painful :)

For that matter, can I just copy <build-server>\vcpkg\installed dir between machines? Is this different to exporting _raw_?

**Comment by jdx-john** [comment]
Looking a bit more, it seems export works similarly to install, you have to specify each port and triplet. I really just want to create an export of everything I have installed in one go - is this not supported?

**Comment by JackBoosY** [comment]
We don't support export muti targets now.
And you should use feature binarycaching, see [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/binarycaching.md).

**Comment by jdx-john** [comment]
@JackBoosY it seems this might be what we want but after reading that page and the more up to date https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md I don't see how to _use_ the cached files. 
My goal is for developers to use vcpkg locally; I read that vcpkg has binary caching on by default, so I checked in `%LOCALAPPDATA%\vcpkg\archives`:

![image](https://user-images.githubusercontent.com/1654627/128485917-5eadad0e-79b6-4f04-97a7-ce8a478f0d67.png)

Looking in archives I don't know what I expected to see, but it is seemingly a load of folders each containing a few archives (I think one per port/triplet combination or something):
![image](https://user-images.githubusercontent.com/1654627/128486211-a6ead0f3-c8e9-4044-b7ea-897f0bd70903.png) 
![image](https://user-images.githubusercontent.com/1654627/128486288-a6aef473-36d0-4c7a-9d8f-edf57f832d10.png)
![image](https://user-images.githubusercontent.com/1654627/128486342-9a7b4b85-181c-477f-8563-f8acb0934949.png)

The one thing the docs don't seem to say is what I _do_ with all this. It's currently on my build machine, and I want to use it on my development machine. This seems to be the [CI -> Developer user story](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/binarycaching.md#ci---developer) but there is no mention how to _consume_ these caches on my dev-box.

I found this blog post which again explains the concept nicely and how to _generate_ caches, but not how to _consume them_. https://devblogs.microsoft.com/cppblog/vcpkg-accelerate-your-team-development-environment-with-binary-caching-and-manifests/

I get the impression it is very trivial but let's say my cache is in `\\<BUILD-SERVER\vcpkg_cache\` what do I do on my dev box to use it?

**Comment by Osyotr** [comment]
> it seems this might be what we want but after reading that page and the more up to date https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md I don't see how to _use_ the cached files.

Well, you use them the same way you generate them.
I.e. pass `--binarysource=\\<BUILD-SERVER\vcpkg_cache\,r` to vcpkg

> The <rw> optional parameter for certain sources controls whether they will be consulted for downloading binaries (read (default), whether on-demand builds will be uploaded to that remote (write), or both (readwrite).

**Comment by jdx-john** [comment]
@Osyotr so I would still need to run `vcpkg install [...] --binarysource=\\<BUILD-SERVER\vcpkg_cache\,r` for each package on my dev-box, and then `vcpkg integrate install` as normal? I had figured I could just _use_ the binary cache as a single blob if I misunderstood that, it would explain my confusion :)

vcpkg on my dev-box will, what, copy all those remote files into my local vcpkg\installed directory then they're used as normal?

**Comment by Osyotr** [comment]
> vcpkg on my dev-box will, what, copy all those remote files into my local vcpkg\installed directory then they're used as normal?

Yep. When you run `vcpkg install my_lib --binarysource=...,r` it will just copy files from remote folder if they exist, otherwise it will build the library.
